mbf_costmap_nav/src/mbf_costmap_nav/free_pose_search.cpp

```diff
@@ -212,64 +223,74 @@ SearchSolution FreePoseSearch::findValidOrientation(const costmap_2d::Costmap2D&
 
 SearchSolution FreePoseSearch::search() const
 {
+  // restart markers
+  if (viz_)
+  {
+    viz_->deleteMarkers();
+  }
+
   const auto costmap2d = costmap_.getCostmap();
 
   // lock costmap so content doesn't change while adding cell costs
   boost::unique_lock<costmap_2d::Costmap2D::mutex_t> lock(*(costmap2d->getMutex()));
 
   std::unordered_set<int> in_queue_or_visited;
+  std::priority_queue<Cell, std::vector<Cell>, decltype(compare_strategy_)> queue(compare_strategy_);
 
   const std::vector<geometry_msgs::Point> footprint =
       safetyPadding(costmap_, config_.use_padded_fp, config_.safety_dist);
 
-  SearchSolution sol;
-  sol.pose.theta = config_.goal.theta;
-  bool outside_or_unknown{ false };
-
-  // initializing queue with the goal cell
-  Cell start;
-  if (costmap2d->worldToMap(config_.goal.x, config_.goal.y, start.x, start.y))
-  {
-    start.cost = costmap2d->getCost(start.x, start.y);
-  }
-  else
+  // enforce bounds if goal is outside the map; and get the pose of the center of the cell
+  Cell goal_cell;
+  geometry_msgs::Pose2D goal_cell_pose;  // this is the pose of the center of the goal cell (not the goal pose)
+  int goal_cell_x, goal_cell_y;
+  costmap2d->worldToMapEnforceBounds(config_.goal.x, config_.goal.y, goal_cell_x, goal_cell_y);
+  goal_cell.x = static_cast<unsigned int>(goal_cell_x);
+  goal_cell.y = static_cast<unsigned int>(goal_cell_y);
+  goal_cell.cost = costmap2d->getCost(goal_cell.x, goal_cell.y);
+  costmap2d->mapToWorld(goal_cell.x, goal_cell.y, goal_cell_pose.x, goal_cell_pose.y);
+
+  unsigned int dummy_x, dummy_y;
+  // don't start the search from the goal pose if goal is not within bounds
+  bool test_goal_pose = costmap2d->worldToMap(config_.goal.x, config_.goal.y, dummy_x, dummy_y);
+
+  // add goal cell to queue if it is within linear tolerance
+  if (std::hypot(goal_cell_pose.x - config_.goal.x, goal_cell_pose.y - config_.goal.y) <= config_.linear_tolerance)
   {
-    // enforce bounds if goal is outside the map
-    // note that distance tolerance is wrt this new start cell
-    int cell_x, cell_y;
-    costmap2d->worldToMapEnforceBounds(config_.goal.x, config_.goal.y, cell_x, cell_y);
-    start.x = static_cast<unsigned int>(cell_x);
-    start.y = static_cast<unsigned int>(cell_y);
-    start.cost = costmap2d->getCost(start.x, start.y);
+    queue.push(goal_cell);
+    in_queue_or_visited.insert(costmap2d->getIndex(goal_cell.x, goal_cell.y));
   }
-  in_queue_or_visited.insert(costmap2d->getIndex(start.x, start.y));
 
-  std::priority_queue<Cell, std::vector<Cell>, decltype(compare_strategy_)> queue(compare_strategy_);
-  queue.push(start);
-
-  // restart markers
-  if (viz_)
-  {
-    viz_->deleteMarkers();
-  }
-
-  while (!queue.empty())
+  SearchSolution sol;
+  sol.pose.theta = config_.goal.theta;
```

Comment:
S: do this inside the while loop. Right now it works under the assumption that the value will never be changed inside the loop, which could change in the future